[documentation] deepsparse user guide (#959)
neuralmagic/deepsparse: user-guide/README.md

# DeepSparse User Guide

This directory demonstrates usage of DeepSparse's key API, including:
- [Benchmarking CLI](#performance-benchmarking)
- [Engine API](#engine)
- [Pipeline API](#pipeline)
- [Server API](#server)

## Installation

Install via `pip`. Using a virtual enviornment is highly recommended.

```bash
pip install deepsparse
```

See the [DeepSparse Installation page](https://docs.neuralmagic.com/get-started/install/deepsparse) for further installation options.

## [Performance Benchmarking](deepsparse-benchmarking.md)

DeepSparse's key feature is its performance on commodity CPUs. DeepSparse is competitive with other CPU runtimes
like ONNX Runtime for unoptimized dense models. However, when optimization techniques like pruning and quantization 

> are applied to a model, DeepSparse can achieve an order-of-magnitude speedup. As an example, let's compare DeepSparse and ORT's performance on BERT. In SparseZoo, there is [90% pruned-quantized BERT](https://sparsezoo.neuralmagic.com/models/nlp%2Fsentiment_analysis%2Fobert-base%2Fpytorch%2Fhuggingface%2Fsst2%2Fpruned90_quant-none). 

Reviewer: ```suggestion
are applied to a model, DeepSparse can achieve an order-of-magnitude speedup. As an example, compare DeepSparse and ORT's performance on BERT. In SparseZoo, there is [90% pruned-quantized BERT](https://sparsezoo.neuralmagic.com/models/nlp%2Fsentiment_analysis%2Fobert-base%2Fpytorch%2Fhuggingface%2Fsst2%2Fpruned90_quant-none). 
```

Running this model on an AWS `c6i.16xlarge` instance, DeepSparse achieves a ***12x speedup*** over ORT!

ORT achieves 18.5 items/second running BERT (make sure you have ORT installed `pip install onnxruntime`):
```bash
deepsparse.benchmark zoo:nlp/text_classification/obert-base/pytorch/huggingface/mnli/base-none -b 64 -s sync -nstreams 1 -i [64,384] -e onnxruntime

>> Original Model Path: zoo:nlp/text_classification/obert-base/pytorch/huggingface/mnli/base-none
>> Batch Size: 64
>> Scenario: sync
>> Throughput (items/sec): 18.5742
```

DeepSparse achieves 226 items/second running the pruned-quantized version of BERT:

```bash
deepsparse.benchmark zoo:nlp/sentiment_analysis/obert-base/pytorch/huggingface/sst2/pruned90_quant-none -b 64 -s sync -nstreams 1 -e onnxruntime

>> Original Model Path: zoo:nlp/text_classification/obert-base/pytorch/huggingface/mnli/pruned90_quant-none
>> Batch Size: 64
>> Scenario: sync
>> Throughput (items/sec): 226.6340
```

**Pro-Tip:** In place of a [SparseZoo](https://sparsezoo.neuralmagic.com/) stubs, you can pass a local ONNX file to test your model.

## Deployment APIs

> Now that we have seen DeepSparse's performance gains, let's take a look at how we can add DeepSparse to an application.

Reviewer: ```suggestion
Now that we have seen DeepSparse's performance gains, we can add DeepSparse to an application.
```

DeepSparse includes three deployment APIs:

> - Engine is the lowest-level API. With Engine, you pass tensors and recieve the raw logits.

Reviewer: ```suggestion
- Engine is the lowest-level API. With Engine, you pass tensors and receive the raw logits.
```

- Pipeline wraps the Engine with pre- and post-processing. With Pipeline, you pass raw data and

> recieve the prediction.

Reviewer: ```suggestion
receive the prediction.
```

- Server wraps Pipelines with a REST API using FastAPI. With Server, you send raw data over HTTP

> and recieve the prediction.

Reviewer: ```suggestion
and receive the prediction.
```

> Let's walk through a simple example of each API to give a sense of usage. As an example, we will use

Reviewer: ```suggestion
The following are simple examples of each API to get a sense of how it is used. For the example, we will use
```

DeepSparse Pipelines wrap Engine with pre- and post-processing (as well as other utilities), 
enabling you to send raw data to DeepSparse and recieve the post-processed prediction.

The example below downloads a 90% pruned-quantized BERT model for sentiment analysis 
in ONNX format from SparseZoo, sets up a pipeline, and runs inference on sample data.

```python
from deepsparse import Pipeline

# download onnx, set up pipeline
zoo_stub = "zoo:nlp/sentiment_analysis/obert-base/pytorch/huggingface/sst2/pruned90_quant-none"  
batch_size = 1
sentiment_analysis_pipeline = Pipeline.create(
  task="sentiment-analysis",    # name of the task
  model_path=zoo_stub,          # zoo stub or path to local onnx file
  batch_size=batch_size         # default is batch 1
)

# run inference (input is a sentence, output is the prediction)
prediction = sentiment_analysis_pipeline("I love using DeepSparse Pipelines")
print(prediction)
# > labels=['positive'] scores=[0.9954759478569031]
```

### [Server](deepsparse-server.md)

Server wraps Pipelines with REST APIs, that make it easy to stand up a model serving endpoint

> running DeepSparse. This enables you to send raw data to DeepSparse over HTTP and recieve the post-processed

Reviewer: ```suggestion
running DeepSparse. This enables you to send raw data to DeepSparse over HTTP and receive the post-processed
```
